extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/BasicRoleBasedAuthorizer.java

```diff
@@ -0,0 +1,142 @@
+import io.druid.server.security.AuthenticationResult;
+import io.druid.server.security.Authorizer;
+import io.druid.server.security.Resource;
+import io.druid.server.security.ResourceAction;
+
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@JsonTypeName("basic")
+public class BasicRoleBasedAuthorizer implements Authorizer
+{
+  private static final int DEFAULT_PERMISSION_CACHE_SIZE = 5000;
+  private final BasicAuthorizerStorageConnector dbConnector;
+  private final LoadingCache<String, Pattern> permissionPatternCache;
+  private final int permissionCacheSize;
+  private final BasicAuthDBConfig dbConfig;
+
+  @JsonCreator
+  public BasicRoleBasedAuthorizer(
+      @JacksonInject BasicAuthorizerStorageConnector dbConnector,
+      @JsonProperty("dbPrefix") String dbPrefix,
+      @JsonProperty("permissionCacheSize") Integer permissionCacheSize
+  )
+  {
+    this.dbConfig = new BasicAuthDBConfig(dbPrefix, null, null);
+    this.dbConnector = dbConnector;
+    this.permissionCacheSize = permissionCacheSize == null ? DEFAULT_PERMISSION_CACHE_SIZE : permissionCacheSize;
+    this.permissionPatternCache = Caffeine.newBuilder()
+                                          .maximumSize(this.permissionCacheSize)
+                                          .build(regexStr -> Pattern.compile(regexStr));
+  }
+
+  /**
+   * constructor for unit tests
+   */
+  public BasicRoleBasedAuthorizer(
+      BasicAuthorizerStorageConnector dbConnector,
+      String dbPrefix,
+      int permissionCacheSize
+  )
+  {
+    this.dbConnector = dbConnector;
+    this.permissionCacheSize = permissionCacheSize;
+    this.permissionPatternCache = Caffeine.newBuilder()
+                                          .maximumSize(this.permissionCacheSize)
+                                          .build(regexStr -> Pattern.compile(regexStr));
+    this.dbConfig = new BasicAuthDBConfig(dbPrefix, null, null);
+  }
+
+  @Override
+  public Access authorize(
+      AuthenticationResult authenticationResult, Resource resource, Action action
+  )
+  {
+    if (authenticationResult == null) {
+      throw new IAE("WTF? authenticationResult should never be null.");
+    }
+
+    List<Map<String, Object>> permissions = dbConnector.getPermissionsForUser(dbConfig.getDbPrefix(), authenticationResult.getIdentity());
+
+    for (Map<String, Object> permission : permissions) {
+      if (permissionCheck(resource, action, permission)) {
+        return new Access(true);
+      }
+    }
+
+    return new Access(false);
+  }
+
+  public BasicAuthorizerStorageConnector getDbConnector()
+  {
+    return dbConnector;
+  }
+
+  public String getDBPrefix()
+  {
+    return dbConfig.getDbPrefix();
+  }
+
+  private boolean permissionCheck(Resource resource, Action action, Map<String, Object> permission)
+  {
+    ResourceAction permissionResourceAction = (ResourceAction) permission.get("resourceAction");
```

Comment:
The cast shouldn't be done here, instead it should already be the correct type when the `permission` object is generated from JSON. This might mean instead of a `Map<String, Object>` it should be a really, real, live object like `class Permission`.